File: clients/ts/src/transport/ws.ts
/**
 * WebSocket implementation of {@link ProtocolTransport}.
 *
 * Wire contract (shared with the reference node server and the Dart/Rust
 * clients — do not deviate):
 * - endpoint `/sync/ws`, JSON text frames, one JSON object per frame.
 * - client→server: `{"v":1,"type":"push"|"pull"|"snapshot","requestId":"<unique>", ...body}`
 * - server→client: `{"v":1,"type":"push-result"|"pull-result"|"snapshot-result","requestId",...body}`,
 *   `{"v":1,"type":"error","requestId","code","message","retryable?"}`, and the
 *   unsolicited pull hint `{"v":1,"type":"changed","watermark":<number>}`.
 *
 * The transport is lazy: it dials on the first request, correlates concurrent
 * requests by `requestId`, and reconnects with full-jitter exponential backoff
 * while the caller's sync loop is the one deciding when to retry a failed
 * request. `changed` hints reach the application through `onChanged`, which
 * should call `ProtocolSyncLoop.hint()` — hints are wake-ups, never data.
 */
import {
  ProtocolTransport,
  ResetRequired,
  SyncTransportError,
  computeRetryDelay,
} from '../sync-loop.js';
import type {
  PullResponse,
  PushRequest,
  PushResponse,
  SnapshotResponse,
} from '../protocol.js';

/** Structural WebSocket so Node (ws / undici), browsers, and tests all fit. */
export interface WebSocketLike {
  readyState: number;
  send(data: string): void;
  close(code?: number, reason?: string): void;
  addEventListener(type: string, listener: (event: any) => void): void;
  removeEventListener?(type: string, listener: (event: any) => void): void;
}

/** Supplies a bearer token for the connection URL. Return null for none. */
export interface AuthTokenProvider {
  getToken(): Promise<string | null> | string | null;
}

export interface WebSocketTransportOptions {
  /** Absolute or origin-relative URL of the `/sync/ws` endpoint. */
  url: string;
  /**
   * Constructor/factory for sockets. Defaults to `globalThis.WebSocket`.
   * Injectable for Node (`ws`) and for tests.
   */
  webSocketFactory?: (url: string) => WebSocketLike;
  /**
   * Pluggable session source (Supabase session, shared-auth, …). The token is
   * appended as a `token` query parameter at dial time, so an expired token
   * picks up its replacement on the next reconnect.
   */
  auth?: AuthTokenProvider;
  /** Reject a dial or in-flight request after this long. Default 20s. */
  requestTimeoutMs?: number;
  /** Called for every unsolicited `changed` frame. Wire to `loop.hint()`. */
  onChanged?: (watermark: number) => void;
  /**
   * Used when the socket cannot be established (or for `snapshot` if the
   * server only serves snapshots over HTTP). Typical value: the HTTP
   * transport. Without a fallback, failures surface as retryable transport
   * errors and the sync loop backs off.
   */
  fallback?: ProtocolTransport;
  reconnectBaseMs?: number;
  reconnectMaxMs?: number;
  random?: () => number;
  now?: () => number;
  setTimeoutFn?: typeof setTimeout;
  clearTimeoutFn?: typeof clearTimeout;
}

/** Host of `url` when its scheme is cleartext `ws://` or `http://`, else null. */
function cleartextHost(url: string): string | null {
  const match = /^(ws|http):\/\//i.exec(url);
  if (!match) return null;
  const authority = url.slice(match[0].length).split(/[/?#]/, 1)[0] ?? '';
  const hostPort = authority.includes('@')
    ? authority.slice(authority.lastIndexOf('@') + 1)
    : authority;
  if (hostPort.startsWith('[')) {
    return hostPort.slice(1, hostPort.indexOf(']')).toLowerCase();
  }
  const colon = hostPort.indexOf(':');
  return (colon === -1 ? hostPort : hostPort.slice(0, colon)).toLowerCase();
}

/** Loopback, private/link-local IPs, and in-cluster names. */
function internalHostAllowed(host: string): boolean {
  if (host === '' || host === 'localhost' || host.endsWith('.localhost')) {
    return true;
  }
  if (host === '::1' || /^f[cd]/.test(host) || /^fe[89ab]/.test(host)) {
    return true;
  }
  const v4 = host.match(/^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/);
  if (v4) {
    const a = Number(v4[1]);
    const b = Number(v4[2]);
    return (
      a === 127 ||
      a === 10 ||
      (a === 172 && b >= 16 && b <= 31) ||
      (a === 192 && b === 168) ||
      (a === 169 && b === 254)
    );
  }
  return (
    !host.includes('.') ||
    host.endsWith('.svc.cluster.local') ||
    host.endsWith('.internal')
  );
}

/**
 * Refuse to dial a public host over cleartext while carrying a token.
 *
 * The token travels in the query string (see `AuthTokenProvider`), so over
 * `ws://` it is readable by anyone on the path — and unlike a header it is also
 * the kind of value that ends up in proxy and server access logs. An
 * origin-relative URL inherits the page's scheme and is left alone.
 */
function requireEncryptedTransport(url: string): void {
  const host = cleartextHost(url);
  if (host !== null && !internalHostAllowed(host)) {
    throw new Error(
      `opto-sync: refusing to send a session token over cleartext to public host "${host}": ` +
        'use wss:// (or https://), an in-cluster address, or loopback',
    );
  }
}

type RequestType = 'push' | 'pull' | 'snapshot';
type ResultFrameType = 'push-result' | 'pull-result' | 'snapshot-result';

const RESULT_FRAME_BY_REQUEST: Readonly<Record<RequestType, ResultFrameType>> = {
  push: 'push-result',
  pull: 'pull-result',
  snapshot: 'snapshot-result',
};

interface PendingRequest {
  generation: number;
  expectedType: ResultFrameType;
  resolve: (frame: Record<string, unknown>) => void;
  reject: (error: Error) => void;
  timer: ReturnType<typeof setTimeout>;
}

interface ConnectedSocket {
  socket: WebSocketLike;
  generation: number;
}

interface ConnectingAttempt {
  generation: number;
  promise: Promise<ConnectedSocket>;
  cancel: (error: SyncTransportError) => void;
}

const OPEN = 1;
const CLOSED = 3;
const MAX_INBOUND_FRAME_BYTES = 1024 * 1024;

function textByteLength(value: string): number {
  return typeof TextEncoder === 'function'
    ? new TextEncoder().encode(value).byteLength
    : value.length;
}

function frameError(frame: Record<string, unknown>): SyncTransportError {
  return new SyncTransportError(
    typeof frame.message === 'string' ? frame.message : 'sync websocket error',
    frame.retryable !== false,
    undefined,
    typeof frame.code === 'string' ? frame.code : 'WS_ERROR',
  );
}

export class WebSocketTransport implements ProtocolTransport {
  private socket: ConnectedSocket | undefined;
  private connecting: ConnectingAttempt | undefined;
  private readonly pending = new Map<string, PendingRequest>();
  private nextRequestId = 0;
  private socketGeneration = 0;
  private consecutiveDialFailures = 0;
  private disposed = false;
  private readonly seed = Math.random().toString(36).slice(2, 10);

  constructor(private readonly options: WebSocketTransportOptions) {
    if (!options.url) throw new RangeError('WebSocketTransport requires a url');
    const timeoutMs = options.requestTimeoutMs ?? 20_000;
    if (!Number.isFinite(timeoutMs) || timeoutMs <= 0) {
      throw new RangeError('requestTimeoutMs must be a positive finite number');
    }
  }

  async push(request: PushRequest, signal: AbortSignal): Promise<PushResponse> {
    const frame = await this.request(
      'push',
      request as unknown as Record<string, unknown>,
      signal,
    );
    return frame as unknown as PushResponse;
  }

  async pull(
    checkpoint: string,
    limit: number,
    signal: AbortSignal,
  ): Promise<PullResponse | ResetRequired> {
    const frame = await this.request('pull', { checkpoint, limit }, signal);
    return frame as unknown as PullResponse | ResetRequired;
  }

  async snapshot(
    signal: AbortSignal,
    reset?: ResetRequired,
  ): Promise<SnapshotResponse> {
    if (this.options.fallback) {
      // Snapshots can be large; the HTTP fallback (CDN-cacheable, resumable)
      // is preferred whenever the caller supplies one.
      return this.options.fallback.snapshot(signal, reset);
    }
    const frame = await this.request('snapshot', {}, signal);
    return frame as unknown as SnapshotResponse;
  }

  /** Closes the socket and rejects all in-flight and dialing requests. */
  dispose(): void {
    if (this.disposed) return;
    this.disposed = true;
    this.socketGeneration += 1;
    const error = new SyncTransportError('transport disposed', false);
    const attempt = this.connecting;
    this.connecting = undefined;
    attempt?.cancel(error);
    this.failPending(error);
    const active = this.socket;
    this.socket = undefined;
    if (active) this.closeSocket(active.socket, 1000, 'dispose');
  }

  private async request(
    type: RequestType,
    body: Record<string, unknown>,
    signal: AbortSignal,
  ): Promise<Record<string, unknown>> {
    if (this.disposed) {
      throw new SyncTransportError('transport disposed', false);
    }
    let connection: ConnectedSocket;
    try {
      connection = await this.connect();
    } catch (error) {
      if (!this.disposed && this.options.fallback) {
        return this.viaFallback(type, body, signal);
      }
      throw error;
    }

    if (!this.isCurrentConnection(connection)) {
      throw new SyncTransportError(
        'websocket closed before request send',
        true,
        undefined,
        'WS_CLOSED',
      );
    }

    const { socket, generation } = connection;
    const requestId = `${this.seed}-${++this.nextRequestId}`;
    const timeoutMs = this.timeoutMs();
    const setTimeoutFn = this.options.setTimeoutFn ?? setTimeout;
    const clearTimeoutFn = this.options.clearTimeoutFn ?? clearTimeout;

    return new Promise<Record<string, unknown>>((resolve, reject) => {
      const abortListener = () => {
        settle();
        const error = new Error('aborted');
        error.name = 'AbortError';
        reject(error);
      };
      const settle = () => {
        const entry = this.pending.get(requestId);
        if (entry) clearTimeoutFn(entry.timer);
        this.pending.delete(requestId);
        signal.removeEventListener('abort', abortListener);
      };
      if (signal.aborted) {
        abortListener();
        return;
      }
      if (!this.isCurrentConnection(connection)) {
        reject(
          new SyncTransportError(
            'websocket closed before request send',
            true,
            undefined,
            'WS_CLOSED',
          ),
        );
        return;
      }
      signal.addEventListener('abort', abortListener, { once: true });
      const timer = setTimeoutFn(() => {
        settle();
        reject(
          new SyncTransportError(
            `websocket ${type} timed out`,
            true,
            undefined,
            'WS_TIMEOUT',
          ),
        );
      }, timeoutMs);
      this.pending.set(requestId, {
        generation,
        expectedType: RESULT_FRAME_BY_REQUEST[type],
        resolve: (frame) => {
          settle();
          resolve(frame);
        },
        reject: (error) => {
          settle();
          reject(error);
        },
        timer,
      });
      try {
        socket.send(JSON.stringify({ v: 1, type, requestId, ...body }));
      } catch {
        const entry = this.pending.get(requestId);
        entry?.reject(
          new SyncTransportError(
            'websocket send failed',
            true,
            undefined,
            'WS_SEND_FAILED',
          ),
        );
      }
    });
  }

  private async viaFallback(
    type: RequestType,
    body: Record<string, unknown>,
    signal: AbortSignal,
  ): Promise<Record<string, unknown>> {
    const fallback = this.options.fallback!;
    if (type === 'push') {
      return (await fallback.push(
        body as unknown as PushRequest,
        signal,
      )) as unknown as Record<string, unknown>;
    }
    if (type === 'pull') {
      return (await fallback.pull(
        String(body.checkpoint),
        Number(body.limit),
        signal,
      )) as unknown as Record<string, unknown>;
    }
    return (await fallback.snapshot(signal)) as unknown as Record<
      string,
      unknown
    >;
  }

  private async connect(): Promise<ConnectedSocket> {
    if (this.disposed) {
      throw new SyncTransportError('transport disposed', false);
    }
    if (this.socket && this.socket.socket.readyState === OPEN) {
      return this.socket;
    }
    if (!this.connecting) {
      const attempt = this.beginDial();
      this.connecting = attempt;
      void attempt.promise.then(
        () => {
          if (this.connecting === attempt) this.connecting = undefined;
        },
        () => {
          if (this.connecting === attempt) this.connecting = undefined;
        },
      );
    }
    return this.connecting.promise;
  }

  private beginDial(): ConnectingAttempt {
    const generation = ++this.socketGeneration;
    let cancel!: (error: SyncTransportError) => void;
    const cancellation = new Promise<ConnectedSocket>((_resolve, reject) => {
      cancel = reject;
    });
    const promise = Promise.race([
      this.dial(generation),
      cancellation,
    ]);
    return { generation, promise, cancel };
  }

  private async dial(generation: number): Promise<ConnectedSocket> {
    let token: string | null | undefined;
    try {
      token = await this.options.auth?.getToken();
    } catch {
      throw this.dialFailure(
        'websocket authentication token lookup failed',
        'WS_AUTH_FAILED',
      );
    }
    if (this.disposed || generation !== this.socketGeneration) {
      throw new SyncTransportError('websocket dial superseded', false);
    }

    let url = this.options.url;
    if (token) {
      requireEncryptedTransport(url);
      if (url.includes('#')) {
        throw new SyncTransportError(
          'websocket URL must not contain a fragment when authentication is enabled',
          false,
          undefined,
          'WS_INVALID_URL',
        );
      }
      const separator = url.includes('?') ? '&' : '?';
      url = `${url}${separator}token=${encodeURIComponent(token)}`;
    }
    const factory =
      this.options.webSocketFactory ??
      ((target: string) => {
        const Ctor = (globalThis as Record<string, unknown>).WebSocket as
          | (new (url: string) => WebSocketLike)
          | undefined;
        if (!Ctor) {
          throw new SyncTransportError(
            'no WebSocket implementation available; pass webSocketFactory',
            false,
            undefined,
            'WS_UNAVAILABLE',
          );
        }
        return new Ctor(target);
      });

    return new Promise<ConnectedSocket>((resolve, reject) => {
      let socket: WebSocketLike;
      try {
        socket = factory(url);
      } catch {
        reject(this.dialFailure('websocket constructor failed'));
        return;
      }

      let opened = false;
      let dialSettled = false;
      const setTimeoutFn = this.options.setTimeoutFn ?? setTimeout;
      const clearTimeoutFn = this.options.clearTimeoutFn ?? clearTimeout;
      const timeout = setTimeoutFn(() => {
        if (dialSettled) return;
        dialSettled = true;
        this.closeSocket(socket, 1001, 'dial timeout');
        reject(this.dialFailure('websocket dial timed out', 'WS_DIAL_TIMEOUT'));
      }, this.timeoutMs());
      (
        timeout as ReturnType<typeof setTimeout> & { unref?: () => void }
      ).unref?.();

      const clearDialTimeout = () => clearTimeoutFn(timeout);
      const onOpen = () => {
        if (dialSettled) return;
        if (this.disposed || generation !== this.socketGeneration) {
          dialSettled = true;
          clearDialTimeout();
          this.closeSocket(socket, 1000, 'superseded dial');
          reject(new SyncTransportError('websocket dial superseded', false));
          return;
        }
        opened = true;
        dialSettled = true;
        clearDialTimeout();
        this.consecutiveDialFailures = 0;
        const connection = { socket, generation };
        this.socket = connection;
        resolve(connection);
      };
      const onMessage = (event: { data: unknown }) => {
        this.onFrame(event.data, socket, generation);
      };
      const onClose = () => {
        const generationCurrent = generation === this.socketGeneration;
        const active = this.isActiveSocket(socket, generation);
        if (active) {
          this.socket = undefined;
          this.failPending(
            new SyncTransportError(
              'websocket closed',
              true,
              undefined,
              'WS_CLOSED',
            ),
            generation,
          );
        }
        if (!opened && !dialSettled && generationCurrent) {
          dialSettled = true;
          clearDialTimeout();
          reject(
            this.dialFailure(
              'websocket closed during dial',
              'WS_DIAL_FAILED',
            ),
          );
        }
      };
      const onError = () => {
        if (generation !== this.socketGeneration) return;
        if (!opened && !dialSettled) {
          dialSettled = true;
          clearDialTimeout();
          this.closeSocket(socket, 1011, 'connection failed');
          reject(
            this.dialFailure(
              'websocket connection failed',
              'WS_DIAL_FAILED',
            ),
          );
          return;
        }
        if (this.isActiveSocket(socket, generation)) {
          this.failConnection(
            socket,
            generation,
            new SyncTransportError(
              'websocket connection error',
              true,
              undefined,
              'WS_SOCKET_ERROR',
            ),
            1011,
            'connection error',
          );
        }
      };

      socket.addEventListener('open', onOpen);
      socket.addEventListener('message', onMessage);
      socket.addEventListener('close', onClose);
      socket.addEventListener('error', onError);
    });
  }

  private dialFailure(
    message: string,
    code = 'WS_DIAL_FAILED',
  ): SyncTransportError {
    this.consecutiveDialFailures += 1;
    const retryAfterMs = computeRetryDelay(
      Math.min(this.consecutiveDialFailures, 20),
      this.options.reconnectBaseMs ?? 500,
      this.options.reconnectMaxMs ?? 30_000,
      this.options.random ?? Math.random,
    );
    return new SyncTransportError(
      message,
      true,
      retryAfterMs,
      code,
    );
  }

  private timeoutMs(): number {
    return this.options.requestTimeoutMs ?? 20_000;
  }

  private failPending(error: SyncTransportError, generation?: number): void {
    for (const entry of [...this.pending.values()]) {
      if (generation === undefined || entry.generation === generation) {
        entry.reject(error);
      }
    }
  }

  private isCurrentConnection(connection: ConnectedSocket): boolean {
    return (
      !this.disposed &&
      connection.generation === this.socketGeneration &&
      this.socket === connection &&
      connection.socket.readyState === OPEN
    );
  }

  private isActiveSocket(socket: WebSocketLike, generation: number): boolean {
    return (
      !this.disposed &&
      generation === this.socketGeneration &&
      this.socket?.generation === generation &&
      this.socket.socket === socket &&
      socket.readyState === OPEN
    );
  }

  private failConnection(
    socket: WebSocketLike,
    generation: number,
    error: SyncTransportError,
    closeCode: number,
    closeReason: string,
  ): void {
    if (!this.isActiveSocket(socket, generation)) return;
    this.socket = undefined;
    this.failPending(error, generation);
    this.closeSocket(socket, closeCode, closeReason);
  }

  private closeSocket(
    socket: WebSocketLike,
    code: number,
    reason: string,
  ): void {
    if (socket.readyState === CLOSED) return;
    try {
      socket.close(code, reason);
    } catch {
      // Generation fencing is the source of truth even if provider close fails.
    }
  }

  private onFrame(
    data: unknown,
    socket: WebSocketLike,
    generation: number,
  ): void {
    if (!this.isActiveSocket(socket, generation)) return;
    if (typeof data !== 'string') {
      this.failConnection(
        socket,
        generation,
        new SyncTransportError(
          'websocket protocol requires text frames',
          false,
          undefined,
          'WS_BINARY_FRAME',
        ),
        1003,
        'text frames required',
      );
      return;
    }
    if (textByteLength(data) > MAX_INBOUND_FRAME_BYTES) {
      this.failConnection(
        socket,
        generation,
        new SyncTransportError(
          'websocket frame exceeds the protocol limit',
          false,
          undefined,
          'WS_FRAME_TOO_LARGE',
        ),
        1009,
        'frame too large',
      );
      return;
    }

    let frame: Record<string, unknown>;
    try {
      frame = JSON.parse(data) as Record<string, unknown>;
    } catch {
      return; // a malformed broadcast must not kill the connection
    }
    if (
      frame === null ||
      typeof frame !== 'object' ||
      Array.isArray(frame) ||
      frame.v !== 1
    ) {
      return;
    }

    if (frame.type === 'changed') {
      if (
        typeof frame.watermark === 'number' &&
        Number.isSafeInteger(frame.watermark) &&
        frame.watermark >= 0
      ) {
        try {
          this.options.onChanged?.(frame.watermark);
        } catch {
          // hints are best-effort; a listener error must not break the socket
        }
      }
      return;
    }

    const requestId = frame.requestId;
    if (typeof requestId !== 'string') return;
    const entry = this.pending.get(requestId);
    if (!entry || entry.generation !== generation) return;

    if (frame.type === 'error') {
      entry.reject(frameError(frame));
      return;
    }
    if (frame.type !== entry.expectedType) {
      if (
        frame.type === 'push-result' ||
        frame.type === 'pull-result' ||
        frame.type === 'snapshot-result'
      ) {
        this.failConnection(
          socket,
          generation,
          new SyncTransportError(
            `websocket response type ${String(frame.type)} does not match ${entry.expectedType}`,
            false,
            undefined,
            'WS_PROTOCOL_MISMATCH',
          ),
          1002,
          'response type mismatch',
        );
      }
      return;
    }

    const { v: _v, type: _t, requestId: _r, ...body } = frame;
    entry.resolve(body);
  }
}

File: clients/rust/src/transport/ws.rs
//! WebSocket implementation of [`ProtocolTransport`] (wire contract v1).
//!
//! Dials `/sync/ws` lazily on the first request, speaks one JSON object per
//! text frame, correlates concurrent in-flight requests by `requestId`, and
//! reconnects on the next request after a close with full-jitter exponential
//! backoff surfaced as `retry_after` on the dial failure. The bearer token
//! (if an [`AuthTokenProvider`] is configured) is appended as a `?token=`
//! query parameter at every dial, so an expired token picks up its
//! replacement on the next reconnect.
//!
//! When a fallback transport is configured
//! ([`WebSocketProtocolTransport::with_fallback`], typically the HTTP
//! transport), requests fall back to it whenever the socket cannot be
//! established, and `snapshot` always prefers it: snapshots can be large and
//! the HTTP endpoint is cacheable and resumable.
//!
//! Only plain `ws://` endpoints are supported; no TLS backend is linked. The
//! e2e stack and the reference node server both speak `ws://`.

use super::{
    decode_push, decode_pull, decode_snapshot, dispatch_frame, erase, lock, push_body, pull_body,
    AuthTokenProvider, ChangedHandler, Core, FrameCodes, FrameSink, Link, PendingMap,
    RandomSource, RequestError, SyncTransportError, DEFAULT_RECONNECT_BASE, DEFAULT_RECONNECT_MAX,
    DEFAULT_REQUEST_TIMEOUT,
};
use crate::protocol::{PushRequest, PushResponse, SnapshotResponse};
use crate::protocol_sync::{ProtocolTransport, PullResult, ResetRequired, TransportFailure};
use serde_json::Map;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{mpsc, Arc};
use std::time::Duration;
use tungstenite::stream::MaybeTlsStream;
use tungstenite::Message;

static WS_CODES: FrameCodes = FrameCodes {
    kind: "websocket",
    error: "WS_ERROR",
    closed: "WS_CLOSED",
    timeout: "WS_TIMEOUT",
    send_failed: "WS_SEND_FAILED",
    dial_failed: "WS_DIAL_FAILED",
    invalid_response: "WS_INVALID_RESPONSE",
    disposed: "WS_DISPOSED",
};

/// How often the socket thread wakes to flush queued outbound frames while
/// blocked waiting for inbound ones.
const POLL_INTERVAL: Duration = Duration::from_millis(25);

pub struct WebSocketTransportOptions {
    /// Absolute `ws://` URL of the `/sync/ws` endpoint.
    pub url: String,
    /// Token source; appended as `?token=<urlencoded>` at every dial.
    pub auth: Option<Arc<dyn AuthTokenProvider>>,
    /// Called for every unsolicited `changed` frame; wire to the sync loop's
    /// wake mechanism. A hint is a wake-up, never data.
    pub on_changed: Option<ChangedHandler>,
    /// Fail an in-flight request after this long (retryable `WS_TIMEOUT`).
    pub request_timeout: Duration,
    pub reconnect_base: Duration,
    pub reconnect_max: Duration,
    /// Jitter source for reconnect backoff; defaults to the OS CSPRNG.
    pub random: Option<RandomSource>,
}

impl WebSocketTransportOptions {
    pub fn new(url: impl Into<String>) -> Self {
        Self {
            url: url.into(),
            auth: None,
            on_changed: None,
            request_timeout: DEFAULT_REQUEST_TIMEOUT,
            reconnect_base: DEFAULT_RECONNECT_BASE,
            reconnect_max: DEFAULT_RECONNECT_MAX,
            random: None,
        }
    }
}

struct WsConfig {
    url: String,
    auth: Option<Arc<dyn AuthTokenProvider>>,
    on_changed: Option<ChangedHandler>,
}

/// WebSocket [`ProtocolTransport`]. Cheap to clone; clones share the same
/// socket, so concurrent requests from multiple threads multiplex over one
/// connection and are correlated by `requestId`.
#[derive(Clone)]
pub struct WebSocketProtocolTransport {
    core: Arc<Core>,
    config: Arc<WsConfig>,
}

impl WebSocketProtocolTransport {
    pub fn new(options: WebSocketTransportOptions) -> Self {
        Self::build(options, None)
    }

    /// Like [`Self::new`], with a fallback transport (typically HTTP) used
    /// when the socket cannot be established, and always preferred for
    /// `snapshot`.
    pub fn with_fallback<T>(options: WebSocketTransportOptions, fallback: T) -> Self
    where
        T: ProtocolTransport + Send + 'static,
        T::Error: std::fmt::Display,
    {
        Self::build(options, Some(erase(fallback)))
    }

    fn build(options: WebSocketTransportOptions, fallback: Option<super::Fallback>) -> Self {
        Self {
            core: Arc::new(Core::new(
                &WS_CODES,
                options.request_timeout,
                options.reconnect_base,
                options.reconnect_max,
                options.random,
                fallback,
            )),
            config: Arc::new(WsConfig {
                url: options.url,
                auth: options.auth,
                on_changed: options.on_changed,
            }),
        }
    }

    /// Close the socket and fail all in-flight requests permanently. Every
    /// clone of this transport is disposed with it.
    pub fn dispose(&self) {
        self.core.dispose();
    }

    fn raw(
        &self,
        frame_type: &str,
        body: Map<String, serde_json::Value>,
    ) -> Result<Map<String, serde_json::Value>, RequestError> {
        let config = Arc::clone(&self.config);
        self.core.request(&move || dial(&config), frame_type, body)
    }
}

impl ProtocolTransport for WebSocketProtocolTransport {
    type Error = SyncTransportError;

    fn push(
        &mut self,
        request: &PushRequest,
    ) -> Result<PushResponse, TransportFailure<Self::Error>> {
        match self.raw("push", push_body(request)) {
            Ok(frame) => decode_push(frame, &WS_CODES),
            Err(RequestError::Dial(failure)) => match &self.core.fallback {
                Some(fallback) => lock(fallback).push(request),
                None => Err(failure),
            },
            Err(other) => Err(other.into_failure()),
        }
    }

    fn pull(
        &mut self,
        checkpoint: &str,
        limit: usize,
    ) -> Result<PullResult, TransportFailure<Self::Error>> {
        match self.raw("pull", pull_body(checkpoint, limit)) {
            Ok(frame) => decode_pull(frame, &WS_CODES),
            Err(RequestError::Dial(failure)) => match &self.core.fallback {
                Some(fallback) => lock(fallback).pull(checkpoint, limit),
                None => Err(failure),
            },
            Err(other) => Err(other.into_failure()),
        }
    }

    fn snapshot(
        &mut self,
        reset: &ResetRequired,
    ) -> Result<SnapshotResponse, TransportFailure<Self::Error>> {
        // Snapshots can be large; the HTTP fallback (cacheable, resumable) is
        // preferred whenever one is configured.
        if let Some(fallback) = &self.core.fallback {
            return lock(fallback).snapshot(reset);
        }
        match self.raw("snapshot", Map::new()) {
            Ok(frame) => decode_snapshot(frame, &WS_CODES),
            Err(error) => Err(error.into_failure()),
        }
    }
}

/* ------------------------------------------------------------------------ */
/* Dialing and the socket thread                                            */
/* ------------------------------------------------------------------------ */

struct WsSink {
    outbound: mpsc::Sender<String>,
    shutdown: Arc<AtomicBool>,
}

impl FrameSink for WsSink {
    fn send_frame(&self, text: String) -> Result<(), String> {
        self.outbound
            .send(text)
            .map_err(|_| "websocket connection is closed".to_string())
    }

    fn close(&self) {
        self.shutdown.store(true, Ordering::SeqCst);
    }
}

fn dial(config: &WsConfig) -> Result<Link, String> {
    let mut url = config.url.clone();
    if let Some(auth) = &config.auth {
        if let Some(token) = auth.token() {
            if !token.is_empty() {
                url.push(if url.contains('?') { '&' } else { '?' });
                url.push_str("token=");
                url.push_str(&encode_query_component(&token));
            }
        }
    }

    let (mut socket, _response) =
        tungstenite::connect(url.as_str()).map_err(|error| error.to_string())?;
    if let MaybeTlsStream::Plain(stream) = socket.get_mut() {
        // The socket thread multiplexes writes and reads, so reads must
        // yield periodically instead of parking the thread forever.
        stream
            .set_read_timeout(Some(POLL_INTERVAL))
            .map_err(|error| error.to_string())?;
    }

    let pending = Arc::new(PendingMap::default());
    let alive = Arc::new(AtomicBool::new(true));
    let shutdown = Arc::new(AtomicBool::new(false));
    let (outbound, outbound_frames) = mpsc::channel::<String>();
    let io = SocketThread {
        socket,
        outbound_frames,
        pending: Arc::clone(&pending),
        alive: Arc::clone(&alive),
        shutdown: Arc::clone(&shutdown),
        on_changed: config.on_changed.clone(),
    };
    std::thread::Builder::new()
        .name("opto-sync-ws".to_string())
        .spawn(move || io.run())
        .map_err(|error| error.to_string())?;

    Ok(Link {
        sink: Arc::new(WsSink { outbound, shutdown }),
        pending,
        alive,
    })
}

struct SocketThread {
    socket: tungstenite::WebSocket<MaybeTlsStream<std::net::TcpStream>>,
    outbound_frames: mpsc::Receiver<String>,
    pending: Arc<PendingMap>,
    alive: Arc<AtomicBool>,
    shutdown: Arc<AtomicBool>,
    on_changed: Option<ChangedHandler>,
}

impl SocketThread {
    /// Single-owner socket loop: flush queued outbound frames, then poll for
    /// one inbound message. One thread doing both sides means no interleaved
    /// writes, ever.
    fn run(mut self) {
        let close_reason: Option<String> = 'io: loop {
            if self.shutdown.load(Ordering::SeqCst) {
                let _ = self.socket.close(None);
                let _ = self.socket.flush();
                break 'io None;
            }
            loop {
                match self.outbound_frames.try_recv() {
                    Ok(text) => {
                        if let Err(error) = self.socket.send(Message::text(text)) {
                            break 'io Some(format!("websocket send failed: {error}"));
                        }
                    }
                    Err(mpsc::TryRecvError::Empty) => break,
                    Err(mpsc::TryRecvError::Disconnected) => {
                        // Every transport handle is gone; nobody can be
                        // waiting on `pending`.
                        let _ = self.socket.close(None);
                        break 'io None;
                    }
                }
            }
            match self.socket.read() {
                Ok(Message::Text(text)) => dispatch_frame(
                    text.as_str(),
                    &self.pending,
                    self.on_changed.as_ref(),
                    &WS_CODES,
                ),
                Ok(Message::Close(_)) => break 'io Some("websocket closed by server".to_string()),
                Ok(_) => {}
                Err(tungstenite::Error::Io(error))
                    if matches!(
                        error.kind(),
                        std::io::ErrorKind::WouldBlock | std::io::ErrorKind::TimedOut
                    ) => {}
                Err(error) => break 'io Some(format!("websocket closed: {error}")),
            }
        };
        self.alive.store(false, Ordering::SeqCst);
        let message = close_reason.unwrap_or_else(|| "websocket transport shut down".to_string());
        self.pending
            .fail_all(&TransportFailure::retryable(SyncTransportError::new(
                WS_CODES.closed,
                message,
            )));
    }
}

/// Percent-encode a query component (RFC 3986 unreserved characters pass
/// through), matching `encodeURIComponent` closely enough for bearer tokens.
fn encode_query_component(value: &str) -> String {
    let mut encoded = String::with_capacity(value.len());
    for byte in value.bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'.' | b'_' | b'~' => {
                encoded.push(byte as char);
            }
            _ => {
                encoded.push('%');
                encoded.push_str(&format!("{byte:02X}"));
            }
        }
    }
    encoded
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn query_component_encoding_is_byte_wise_percent_encoding() {
        assert_eq!(encode_query_component("plain-token_1.2~3"), "plain-token_1.2~3");
        assert_eq!(encode_query_component("a b+c/d&e=f"), "a%20b%2Bc%2Fd%26e%3Df");
        assert_eq!(encode_query_component("ünï"), "%C3%BCn%C3%AF");
    }
}
